### PR TITLE
Music

### PR DESCRIPTION
**Pull request by Marvi-Marv**
Hey Paul, in case you like the s3 music, you may want my implementation :) thanks for your contribution to the project

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -94,6 +94,17 @@ subprojects {
             implementation 'net.sourceforge.streamsupport:streamsupport-flow:' + rootProject.ext.streamSupportVersion
             implementation 'net.sourceforge.streamsupport:streamsupport-literal:' + rootProject.ext.streamSupportVersion
 
+            /*add music dependencies
```
Comment (paulwedeck) [inline build.gradle:97]: These dependencies are only relevent for swing. So this belongs in go.graphics.swing

File: jsettlers.common/src/main/java/jsettlers/common/music/MusicManager.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+		if (lookupPath.getName().equals(MusicManager.HISTORY_EDITION_MUSIC_FOLDER_NAME)) {
+			MusicManager.fileType = MusicManager.HISTORY_EDITION_FILE_TYPE;
+			MusicManager.musicSet = MusicManager.HISTORY_EDITION_MUSIC_SET;
+			valid = true;
+		} else if (lookupPath.getName().equals(MusicManager.ULTIMATE_EDITION_MUSIC_FOLDER_NAME)) {
+			MusicManager.fileType = MusicManager.ULTIMATE_EDITION_FILE_TYPE;
+			MusicManager.musicSet = MusicManager.ULTIMATE_EDITION_MUSIC_SET;
+			valid = true;
+		} else {
+			// throw new Exception("invalid music folder");
```
Comment (paulwedeck) [inline jsettlers.common/src/main/java/jsettlers/common/music/MusicManager.java:63]: In this case we might aswell just play all the existing files that we can interpret

File: jsettlers.common/src/main/java/jsettlers/common/music/MusicManager.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+		}
+
+		if (valid) {
+			MusicManager.lookupPath = lookupPath;
+		}
+	}
+
+	public static void startMusicThread(final ECivilisation civilisation, final boolean playAll) {
+		if (MusicManager.lookupPath != null) {
```
Comment (paulwedeck) [inline jsettlers.common/src/main/java/jsettlers/common/music/MusicManager.java:72]: This should really be a member function.

File: jsettlers.common/src/main/java/jsettlers/common/music/MusicManager.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	public static void setMusicVolume(final int volume) {
+		if (MusicManager.lookupPath != null) {
+			musicPlayer.setVolume(volume);
+		}
+	}
+
+	private static String[] assembleMusicSet(final ECivilisation civilisation, final boolean playAll) {
+		List<String> list = new ArrayList<String>();
```
Comment (paulwedeck) [inline jsettlers.common/src/main/java/jsettlers/common/music/MusicManager.java:97]: Why not just create Files here?

File: jsettlers.common/src/main/java/jsettlers/common/music/MusicPlayer.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+ *
+ * Supports .mp3 and .ogg music file streams
+ *
+ * dependencies for .ogg and .mp3 files: ogg libraries: vorbisspi1.0.2.jar, tritonus_share.jar, jorbis-0.0.15.jar, jogg-0.0.7.jar mp3 libraries: jl1.0.1.jar, tritonus_share.jar, mp3spi1.9.5.jar
+ *
+ * gradle takes care of dependencies
+ *
+ * @author MarviMarv
+ */
+public final class MusicPlayer {
```
Comment (paulwedeck) [inline jsettlers.common/src/main/java/jsettlers/common/music/MusicPlayer.java:36]: This class is basically a duplicate of SwingSoundPlayer except that the latter takes offsets into s3 sound file and this one takes ordinary sound files.

You might aswell merge the two.

File: jsettlers.common/src/main/java/jsettlers/common/music/MusicThread.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	private final MusicPlayer musicPlayer;
+	private final int startVolume;
+
+	private final String musicFilePath;
+	private final String musicFileType;
+	private final String[] musicTracks;
+
+	private boolean canceled;
+
+	public MusicThread(final MusicPlayer musicPlayer, final int volume, final String musicFilePath, final String musicFileType, final String[] musicTracks) {
```
Comment (paulwedeck) [inline jsettlers.common/src/main/java/jsettlers/common/music/MusicThread.java:20]: You might aswell just merge this class and MusicManager.

File: jsettlers.logic/src/main/java/jsettlers/input/GuiInterface.java [hunk 7]
```diff
@@ -150,6 +151,8 @@ public void run() {
 		}
 		clock.setTaskExecutor(new GuiTaskExecutor(grid, this, this.playerId));
 		connector.addListener(this);
+
+		MusicManager.startMusicThread(player.getCivilisation(), false);
```
Comment (paulwedeck) [inline jsettlers.logic/src/main/java/jsettlers/input/GuiInterface.java:155]: This does not belong here. I would rather put this into MapContent

File: jsettlers.logic/src/main/java/jsettlers/input/GuiInterface.java [hunk 8]
```diff
@@ -385,6 +388,7 @@ public void action(IAction action) {
 				break;
 
 			case EXIT:
+				MusicManager.stopMusicThread();
```
Comment (paulwedeck) [inline jsettlers.logic/src/main/java/jsettlers/input/GuiInterface.java:391]: This should go into MapContent#stop()

**Review by paulwedeck — CHANGES_REQUESTED** [review]
Thank you for writing this pr.
I would like to have the original music in jsettlers but I didn't find the time yet.
Also I only have the gold edition and that saves the music on the cd itself as audio tracks, so the only real way for me to test your pr is if you allow custom sound folders too.
I think you should look at extending SoundPlayer and mirror the structure of SoundManager and BackgroundSound, so that your code fits better into the overall project.

**Comment by Marvi-Marv** [comment]
> Thank you for writing this pr.
> I would like to have the original music in jsettlers but I didn't find the time yet.
> Also I only have the gold edition and that saves the music on the cd itself as audio tracks, so the only real way for me to test your pr is if you allow custom sound folders too.
> I think you should look at extending SoundPlayer and mirror the structure of SoundManager and BackgroundSound, so that your code fits better into the overall project.

Thank you for your feedback Paul. I will try to the code as much as possible in order to have a better fit for the overall project. You could create a new folder ("MUSIC" or "Theme") in your s3 install directory and copy the cd audio files there.
I am currently working on correcting the building images and ressource requirements for all civilizations.

**Comment by paulwedeck** [comment]
> I am currently working on correcting the building images and ressource requirements for all civilizations.
Great to hear that.
If you haven't noticed already: The buildings are mostly defined in xml files in jsettlers.common/src/main/resources/jsettlers/common/buildings/
Also the BuildingCreatorApp in jsettlers.buildingcreator should be helpful.
There are also some tests that validate these xml files.

**Comment by Marvi-Marv** [comment]
> > I am currently working on correcting the building images and ressource requirements for all civilizations.
> 
> Great to hear that.
> If you haven't noticed already: The buildings are mostly defined in xml files in jsettlers.common/src/main/resources/jsettlers/common/buildings/
> Also the BuildingCreatorApp in jsettlers.buildingcreator should be helpful.
> There are also some tests that validate these xml files.

Yes, i have already found these. And the tools are very helpful. I will push the first changes to my repository branch "AssignCorrectBuildingRessources" later this evening. I am still not entirley sure how i can transform the winegrower to the brewery for instance, but i am fine with the images and the ressources. Do you use some kind of messenger or voice tool in order to communicate?

**Comment by paulwedeck** [comment]
You should look at the jobs for the backery. They should be much closer to the Brewery than the vineyard is.
You also shouldn't "transform" one kind of building to another but create a new entry in EBuildingType, etc. for it.
Its not an issue if there isn't a vineyard xml for the egyptians or a brewery for the romans.
We have a discord server. Its at the bottom of the readme. https://discord.gg/2hVV4u6

**Comment by Marvi-Marv** [comment]
added gui and moved files to fit project structure

**Pull request closed without merging**